src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -48,19 +49,51 @@ internal class CodeActionsHandler : ILspServiceDocumentRequestHandler<LSP.CodeAc
             _globalOptions = globalOptions;
         }
 
-        public TextDocumentIdentifier GetTextDocumentIdentifier(CodeActionParams request) => request.TextDocument;
+        public TextDocumentIdentifier GetTextDocumentIdentifier(CodeActionParamsWithOptions request) => request.TextDocument;
 
-        public async Task<LSP.CodeAction[]> HandleRequestAsync(LSP.CodeActionParams request, RequestContext context, CancellationToken cancellationToken)
+        public async Task<LSP.CodeAction[]> HandleRequestAsync(CodeActionParamsWithOptions request, RequestContext context, CancellationToken cancellationToken)
         {
             var document = context.Document;
             Contract.ThrowIfNull(document);
 
-            var options = _globalOptions.GetCodeActionOptionsProvider();
+            var options = new LSPCodeActionOptionsProvider(request.AllowGenerateInHiddenCode, _globalOptions.GetCodeActionOptionsProvider());
 
             var codeActions = await CodeActionHelpers.GetVSCodeActionsAsync(
                 request, document, options, _codeFixService, _codeRefactoringService, cancellationToken).ConfigureAwait(false);
 
             return codeActions;
         }
+
+        private class LSPCodeActionOptionsProvider : AbstractCodeActionOptionsProvider
+        {
+            private readonly bool _allowGenerateInHiddenCode;
+            private readonly CodeActionOptionsProvider _underlyingProvider;
+            private CodeAnalysis.CodeActions.CodeActionOptions? _options;
+
+            public LSPCodeActionOptionsProvider(bool allowGenerateInHiddenCode, CodeActionOptionsProvider underlyingOptions)
+            {
+                _allowGenerateInHiddenCode = allowGenerateInHiddenCode;
+                _underlyingProvider = underlyingOptions;
+            }
+
+            public override CodeAnalysis.CodeActions.CodeActionOptions GetOptions(LanguageServices languageServices)
+            {
+                if (_options is not null)
```

Comment:
Would this return incorrect options if we are first asked for C# options and then for VB options?

Reply:
Yes, it would. Thank you for the extra context above. I'll remove the caching

Reply:
If it is important to cache for perf reasons you can use `ImmutableDictionary<string, CodeActionOptions>` to store them.

Reply:
Yeah, `CodeActionOptionsStorage.GetCodeActionOptions` caches, so I figured it was important enough, I just didn't read the code closely enough to see that it was per language. I'll copy that pattern